Cabal/src/Distribution/Simple/Command.hs

```diff
@@ -663,7 +664,8 @@ commandsRun globalCommand commands args =
 
     callExternal :: a -> String -> [String] -> IO (CommandParse (a, CommandParse action))
     callExternal flags exec cmdArgs = do
-      result <- try $ callProcess exec cmdArgs
+      execPath <- getExecutablePath
+      result <- try $ createProcess (proc exec cmdArgs){env = Just [("CABAL", execPath)]}
```

Comment:
Closer.. but setting `env` like this **overrides** the whole environment, you need to get the current environment first using `getEnvironment`.